File: internal/selector/selector.go
// Package selector implements keep-at's two central decisions: which torrent
// is most urgently in need of seeding, and whether swapping to it right now
// would risk a cascade (every keep-at node globally piling onto the same
// under-seeded torrent at once).
package selector

import (
	"math"
	"sort"

	"github.com/anacrolix/torrent/metainfo"
)

// Candidate is a torrent keep-at is considering downloading, combining
// catalog metadata with a fresh tracker scrape.
type Candidate struct {
	InfoHash  metainfo.Hash
	Title     string
	SizeBytes int64
	Seeders   int
	Leechers  int
	// KeepAtPeers is how many other keep-at nodes were observed in the
	// swarm, gathered by probing. It feeds network-status reporting and is
	// logged as metadata; it deliberately does NOT drive the selection
	// decision - keep-at exists to seed minimally-seeded torrents, so the
	// gate below is keyed on total Seeders, not on how many keep-at nodes
	// happen to be present.
	KeepAtPeers int
	// SeederFloor is the p10 (10th percentile) number of seeders across all
	// torrents in the catalog with at least one seeder, as measured by the
	// most recently completed scan. It anchors SelectionChance: a torrent is
	// judged against how the rest of the catalog is doing, not against a
	// fixed baseline of one. Before any scan has completed it's 0, which
	// makes SelectionChance fall back to the original
	// aggressiveness^(seeders-1) behavior. See SelectionChance.
	SeederFloor int
}

// Available reports whether a candidate meets the availability bar: at
// least one live seed. The plan also allows "peers with sufficient data"
// (i.e. leechers whose combined pieces cover 100% of the torrent even with
// zero full seeds) to count as available, but verifying that requires
// per-peer piece-map inspection across the whole swarm, which is expensive
// to do for every catalog candidate during a scan. keep-at uses the cheaper,
// conservative signal instead: a torrent isn't a candidate until someone
// can serve 100% of it outright.
func (c Candidate) Available() bool {
	return c.Seeders >= 1
}

// Held is a torrent keep-at currently stores and seeds.
type Held struct {
	InfoHash  metainfo.Hash
	Title     string
	SizeBytes int64
	Seeders   int
}

// RankCandidates orders candidates by seeding urgency: fewest seeds first,
// with smaller torrents preferred as a tie-break (they're cheaper to try
// first while keep-at works down the list). Candidates that aren't Available
// are excluded entirely, since they'd stall a download indefinitely.
//
// ramBound, when true, means keep-at is already at its RAM-driven torrent
// cap, so the binding constraint is the per-torrent RAM slot rather than
// disk. The size tie-break then flips to prefer LARGER torrents, because each
// RAM slot seeds a fixed overhead regardless of size - a big torrent fills a
// slot far more efficiently than a small one, which is exactly what a
// RAM-constrained, disk-rich host (e.g. a 1 GB Pi on a 20 TB disk) wants.
func RankCandidates(candidates []Candidate, ramBound bool) []Candidate {
	ranked := make([]Candidate, 0, len(candidates))
	for _, c := range candidates {
		if c.Available() {
			ranked = append(ranked, c)
		}
	}
	sort.SliceStable(ranked, func(i, j int) bool {
		if ranked[i].Seeders != ranked[j].Seeders {
			return ranked[i].Seeders < ranked[j].Seeders
		}
		if ramBound {
			return ranked[i].SizeBytes > ranked[j].SizeBytes
		}
		return ranked[i].SizeBytes < ranked[j].SizeBytes
	})
	return ranked
}

// SelectionChance is n: the probability keep-at proceeds with a candidate
// given how many seeders it already has, relative to how the rest of the
// catalog is doing. See DESIGN.md for the full rationale.
//
// keep-at's purpose is to seed minimally-seeded torrents - not to put a
// keep-at copy on everything. The exponent is the candidate's seeder count
// minus the catalog's p10 seeder floor (seederFloor), floored at zero:
//
//	n = aggressiveness ^ max(0, seeders - seederFloor)
//
// With seederFloor 1 (no completed scan yet, or a catalog where a tenth of
// seeded torrents have just one seeder), this reduces to the original
// n = aggressiveness ^ (seeders - 1): a single-seeder torrent is the
// primary target and passes confidently, and n shrinks toward zero as
// seeders accumulate. As the catalog's overall health improves - the p10
// floor rises - so does the seeder count a candidate must beat to be
// confidently selected. That's deliberate: keep-at's job scales with how
// under-seeded the catalog is, and anchoring to a moving floor means it
// keeps finding genuinely under-seeded content as overall health improves,
// while never assuming health has risen above the measured floor before it
// actually has. A torrent at or below the floor is always selected with
// full confidence.
//
// Note this is deliberately keyed on TOTAL seeders, not on how many other
// keep-at nodes are in the swarm. The keep-at peer count is network-status
// metadata (see Candidate.KeepAtPeers); it doesn't gate selection.
func SelectionChance(aggressiveness float64, seeders int, seederFloor int) float64 {
	if seeders < 1 {
		seeders = 1
	}
	if seederFloor < 1 {
		seederFloor = 1
	}
	exponent := seeders - seederFloor
	if exponent < 0 {
		exponent = 0
	}
	return math.Pow(aggressiveness, float64(exponent))
}

// SeederFloor computes the p10 (10th percentile) number of seeders across
// the given torrents' seeder counts, considering only torrents with at
// least one seeder. It's the anchor x for SelectionChance: a candidate at
// or below the floor passes with full confidence, and confidence decays as
// its seeder count rises above the floor.
//
// The nearest-rank method is used: for n positive counts, the floor is the
// value at rank ceil(0.10 * n) when sorted ascending (i.e. the smallest
// value such that at least 10% of seeded torrents have that many or fewer
// seeders). With no positive counts at all - a catalog where nothing is
// seeded - it returns 0, meaning "no health signal", which SelectionChance
// treats as the conservative single-seeder baseline.
func SeederFloor(seedCounts []int) int {
	var seeded []int
	for _, s := range seedCounts {
		if s > 0 {
			seeded = append(seeded, s)
		}
	}
	if len(seeded) == 0 {
		return 0
	}
	sort.Ints(seeded)
	rank := (len(seeded) + 9) / 10 // ceil(n/10), nearest-rank p10
	idx := rank - 1
	if idx < 0 {
		idx = 0
	}
	return seeded[idx]
}

// MeetsSeedMargin reports whether candidateSeeders is at least minSeedMargin
// lower than every torrent in displaced. It's cheap (no swarm probing
// involved) and deliberately checkable before EvaluateSwap so callers can
// skip an expensive keep-at-peer probe for candidates that would fail this
// check anyway. An empty displaced list always passes: there's nothing to
// beat when keep-at is just filling free space, not swapping.
func MeetsSeedMargin(candidateSeeders int, displaced []Held, minSeedMargin int) bool {
	if len(displaced) == 0 {
		return true
	}
	minDisplacedSeeders := displaced[0].Seeders
	for _, d := range displaced[1:] {
		if d.Seeders < minDisplacedSeeders {
			minDisplacedSeeders = d.Seeders
		}
	}
	return candidateSeeders <= minDisplacedSeeders-minSeedMargin
}

// SwapDecision is the outcome of evaluating whether to swap to a candidate.
type SwapDecision struct {
	ShouldSwap bool
	Chance     float64 // n
	Roll       float64
	Reason     string
}

// ReasonSeedScarcityRollFailed is the Reason on a SwapDecision when the
// seed-scarcity roll did not pass: the candidate already has enough seeders
// relative to the catalog's floor, so keep-at backs off. This is the routine
// outcome for well-seeded candidates - the expected case, not an error - and
// the engine deliberately does not log it (see logEvaluatedCandidate).
const ReasonSeedScarcityRollFailed = "seed-scarcity roll failed; candidate already has enough seeders"

// SeedScarcityBlocked reports whether this decision failed specifically
// because the seed-scarcity roll didn't pass, as opposed to some other
// reason (unavailable candidate, seed margin not met, RAM cap, etc.).
func (d SwapDecision) SeedScarcityBlocked() bool {
	return d.Reason == ReasonSeedScarcityRollFailed
}

// EvaluateSwap decides whether to start downloading candidate, optionally
// displacing the torrents in displaced to make room. It swaps when roll <
// n: see DESIGN.md for why that comparison direction is what actually
// keeps keep-at from wasting slots on torrents that are already healthy.
//
// The gate (n) is keyed on the candidate's TOTAL seeders: a torrent with
// one seed is keep-at's primary target and passes confidently, while a
// torrent with many seeds - already healthy on its own - is effectively
// never selected. The keep-at peer count (Candidate.KeepAtPeers) does not
// gate selection; it's network-status metadata only.
//
// roll must be a fresh uniform [0, 1) draw per call; it's a parameter
// rather than generated internally so this stays deterministic to test.
func EvaluateSwap(candidate Candidate, displaced []Held, minSeedMargin int, aggressiveness float64, roll float64) SwapDecision {
	if !candidate.Available() {
		return SwapDecision{ShouldSwap: false, Reason: "candidate has no live seed"}
	}

	if !MeetsSeedMargin(candidate.Seeders, displaced, minSeedMargin) {
		return SwapDecision{ShouldSwap: false, Reason: "candidate does not beat displaced torrents by the required margin"}
	}

	chance := SelectionChance(aggressiveness, candidate.Seeders, candidate.SeederFloor)
	shouldSwap := roll < chance

	reason := "seed-scarcity roll succeeded"
	if !shouldSwap {
		reason = ReasonSeedScarcityRollFailed
	}

	return SwapDecision{ShouldSwap: shouldSwap, Chance: chance, Roll: roll, Reason: reason}
}

File: src/selector.rs
//! Selection logic: which torrent most urgently needs seeding, and
//! whether swapping to it risks a cascade.

use rand::Rng;

/// A torrent keep-at is considering downloading.
#[derive(Debug, Clone)]
pub struct Candidate {
    pub info_hash: [u8; 20],
    pub title: String,
    pub size_bytes: u64,
    /// Piece count: the RAM-cost axis (rqbit bookkeeping scales ~32 B/piece).
    /// Defaults to 0 (unknown) when the caller has no metainfo handy; a
    /// 0-piece candidate prices like a 1-piece one (never free).
    pub piece_count: u32,
    pub seeders: u32,
    pub leechers: u32,
    /// p10 seeder floor from the last completed scan. 0 (no completed scan)
    /// falls back to the original aggressiveness^(seeders-1) behavior.
    pub seeder_floor: u32,
}

impl Candidate {
    /// Availability bar: at least one live seed.
    pub fn available(&self) -> bool {
        self.seeders >= 1
    }
}

/// A torrent keep-at currently holds and seeds.
#[derive(Debug, Clone)]
pub struct Held {
    pub info_hash: [u8; 20],
    pub title: String,
    pub size_bytes: u64,
    /// Piece count (0 = unknown, e.g. state written before this was tracked).
    pub piece_count: u32,
    pub seeders: u32,
}

/// Order candidates by seeding urgency: fewest seeds first; unavailable
/// (zero-seed) candidates excluded. The size-bias step below never compares
/// across seeder counts — it only orders within the same count.
///
/// Size tie-break is a bounded adaptive bias driven by the host's RAM:disk
/// ratio (`size_bias > 0` favors larger torrents, `< 0` favors smaller ones,
/// `0` keeps prior behavior). The bias is a *tie-break*: a candidate with
/// fewer seeders always outranks one with more, so poorly-seeded torrents
/// keep absolute priority.
///
/// - RAM-bound hosts (little RAM, big disk) get `size_bias > 0`: ties break
///   by bytes-per-RAM-byte, filling scarce RAM slots with the most bytes.
/// - RAM-rich hosts get `size_bias < 0`: ties break smallest-first, spending
///   plentiful slots on the small torrents big-disk hosts skip.
///
/// Piece count stays the RAM axis (rqbit bookkeeping scales ~32 B/piece);
/// byte size is the disk axis. See `size_bias_for_ratio` for how the ratio
/// maps to the exponent.
pub fn rank_candidates(
    mut candidates: Vec<Candidate>,
    size_bias: f64,
    peer_limit: usize,
) -> Vec<Candidate> {
    candidates.retain(|c| c.available());
    candidates.sort_by(|a, b| {
        // Primary key: fewest seeders first — the bias below can never
        // promote across seeder bands. Secondary key: size-bias score,
        // higher first.
        match a.seeders.cmp(&b.seeders) {
            std::cmp::Ordering::Equal => size_bias_score(b, size_bias, peer_limit)
                .partial_cmp(&size_bias_score(a, size_bias, peer_limit))
                .unwrap_or_else(|| a.size_bytes.cmp(&b.size_bytes)),
            ord => ord,
        }
    });
    candidates
}

/// Adjusted bytes-per-RAM-byte score for the size-bias tie-break.
/// `size_bias` is small (|bias| <= 1). Higher score = ranked first.
/// Positive bias rewards raw ratio (larger torrents win ties); zero bias
/// ranks by plain size, legacy order; negative bias ranks by smallest size
/// first (small torrents win ties), with the ratio breaking size ties
/// toward fewer pieces.
fn size_bias_score(c: &Candidate, size_bias: f64, peer_limit: usize) -> f64 {
    // Higher score = ranked first. Positive bias rewards the bytes-per-RAM
    // ratio (larger torrents win ties); zero or negative bias ranks
    // smallest-first (legacy order), with the ratio breaking exact size
    // ties toward fewer pieces.
    let ram = crate::engine::ram::torrent_ram(c.piece_count, peer_limit).max(1) as f64;
    let ratio = (c.size_bytes.max(1) as f64) / ram;
    if size_bias > 0.0 {
        return ratio.powf(size_bias);
    }
    -(c.size_bytes as f64) + ratio / (1.0 + c.size_bytes as f64)
}

/// Eviction-order score for a held torrent: ascending order evicts the
/// lowest score first. This is the same adjusted ratio as
/// [`size_bias_score`] (public so the engine's displaceable-set ordering
/// stays identical to ranking by construction, not by duplication).
/// Zero bias preserves the legacy order (largest seeders evicted first is
/// handled by the caller only when bias is zero — see below).
pub fn eviction_score(c: &Candidate, size_bias: f64, peer_limit: usize) -> f64 {
    if size_bias == 0.0 {
        // Legacy swap path evicted highest-seeded qualifying first; with no
        // bias there is no size signal, so keep that order by returning the
        // seeder count as the score. Callers pass seeders through.
        return c.seeders as f64;
    }
    size_bias_score(c, size_bias, peer_limit)
}

/// n: the probability keep-at proceeds with a candidate given its seeder
/// count relative to the catalog's p10 floor:
/// n = aggressiveness ^ max(0, seeders - floor).
pub fn selection_chance(aggressiveness: f64, seeders: u32, seeder_floor: u32) -> f64 {
    let seeders = seeders.max(1);
    let floor = seeder_floor.max(1);
    let exponent = seeders.saturating_sub(floor);
    aggressiveness.powi(exponent as i32)
}

/// Nearest-rank p10 over positive seeder counts; 0 when nothing is seeded.
pub fn seeder_floor(seed_counts: &[u32]) -> u32 {
    let mut seeded: Vec<u32> = seed_counts.iter().copied().filter(|&s| s > 0).collect();
    if seeded.is_empty() {
        return 0;
    }
    seeded.sort_unstable();
    let rank = seeded.len().div_ceil(10); // ceil(n/10)
    seeded[rank.saturating_sub(1)]
}

/// Whether the candidate beats every displaced torrent by the margin.
/// Empty displaced list always passes (free-space fill, nothing to beat).
pub fn meets_seed_margin(candidate_seeders: u32, displaced: &[Held], min_seed_margin: i32) -> bool {
    if displaced.is_empty() {
        return true;
    }
    let min_displaced = displaced.iter().map(|d| d.seeders).min().unwrap_or(0);
    (candidate_seeders as i64) <= (min_displaced as i64) - (min_seed_margin as i64)
}

#[derive(Debug, Clone)]
pub struct SwapDecision {
    pub should_swap: bool,
    pub chance: f64,
    pub roll: f64,
    pub reason: String,
}

pub const REASON_SEED_SCARCITY_ROLL_FAILED: &str =
    "seed-scarcity roll failed; candidate already has enough seeders";

impl SwapDecision {
    pub fn seed_scarcity_blocked(&self) -> bool {
        self.reason == REASON_SEED_SCARCITY_ROLL_FAILED
    }
}

/// Decide whether to download `candidate`, optionally displacing `displaced`.
/// Swaps when roll < n. `roll` is a caller-supplied uniform [0,1) draw so
/// this stays deterministic to test.
pub fn evaluate_swap(
    candidate: &Candidate,
    displaced: &[Held],
    min_seed_margin: i32,
    aggressiveness: f64,
    roll: f64,
) -> SwapDecision {
    if !candidate.available() {
        return SwapDecision {
            should_swap: false,
            chance: 0.0,
            roll,
            reason: "candidate has no live seed".to_string(),
        };
    }
    if !meets_seed_margin(candidate.seeders, displaced, min_seed_margin) {
        return SwapDecision {
            should_swap: false,
            chance: 0.0,
            roll,
            reason: "candidate does not beat displaced torrents by the required margin".to_string(),
        };
    }
    let chance = selection_chance(aggressiveness, candidate.seeders, candidate.seeder_floor);
    let should_swap = roll < chance;
    SwapDecision {
        should_swap,
        chance,
        roll,
        reason: if should_swap {
            "seed-scarcity roll succeeded".to_string()
        } else {
            REASON_SEED_SCARCITY_ROLL_FAILED.to_string()
        },
    }
}

/// Weighted choice of storage location index by free-space weights.
/// `free[i]` must be >= `need[i]` to be eligible. Deterministic in `roll`.
pub fn choose_location(free: &[u64], need: &[u64], roll: f64) -> Option<usize> {
    let mut eligible: Vec<(usize, u64)> = free
        .iter()
        .zip(need.iter())
        .enumerate()
        .filter(|(_, (&f, &n))| f >= n)
        .map(|(i, (&f, _))| (i, f))
        .collect();
    if eligible.is_empty() {
        return None;
    }
    eligible.sort_by_key(|(i, _)| *i);
    let total: f64 = eligible.iter().map(|(_, f)| *f as f64).sum();
    let mut target = roll * total;
    for (i, f) in &eligible {
        target -= *f as f64;
        if target < 0.0 {
            return Some(*i);
        }
    }
    // Float rounding: fall back to the last candidate.
    eligible.last().map(|(i, _)| *i)
}

/// Draw a uniform [0,1) roll.
pub fn roll(rng: &mut impl Rng) -> f64 {
    rng.gen_range(0.0..1.0)
}

#[cfg(test)]
mod tests {
    use super::*;

    fn cand(seeders: u32, size: u64) -> Candidate {
        Candidate {
            info_hash: [0u8; 20],
            title: String::new(),
            size_bytes: size,
            piece_count: 0,
            seeders,
            leechers: 0,
            seeder_floor: 0,
        }
    }

    #[test]
    fn rank_fewest_seeds_first() {
        // Seeder count dominates: bias never promotes across bands.
        let v = rank_candidates(
            vec![cand(5, 10), cand(1, 100), cand(0, 1), cand(1, 5)],
            1.0,
            8,
        );
        assert_eq!(v.len(), 3);
        assert_eq!(v[0].seeders, 1);
        assert_eq!(v[1].seeders, 1);
        assert_eq!(v[2].seeders, 5);
        // Same bias, balanced order: zero bias keeps legacy size order.
        let v = rank_candidates(
            vec![cand(5, 10), cand(1, 100), cand(0, 1), cand(1, 5)],
            0.0,
            8,
        );
        assert_eq!(v.len(), 3);
        assert_eq!(v[0].size_bytes, 5);
        assert_eq!(v[1].size_bytes, 100);
        assert_eq!(v[2].seeders, 5);
    }

    #[test]
    fn rank_bias_directions() {
        // Positive bias favors larger within a seeder band...
        let v = rank_candidates(vec![cand(1, 5), cand(1, 100)], 1.0, 8);
        assert_eq!(v[0].size_bytes, 100);
        // ...negative bias favors smaller...
        let v = rank_candidates(vec![cand(1, 5), cand(1, 100)], -1.0, 8);
        assert_eq!(v[0].size_bytes, 5);
        // ...and zero bias keeps legacy order (smaller first).
        let v = rank_candidates(vec![cand(1, 5), cand(1, 100)], 0.0, 8);
        assert_eq!(v[0].size_bytes, 5);
    }

    #[test]
    fn rank_ram_bound_prices_pieces_not_bytes() {
        // Same 2 GiB size, different piece counts: fewer pieces wins ties
        // under positive bias (same disk, less RAM).
        let mut a = cand(1, 2 << 30);
        a.piece_count = 100;
        let mut b = cand(1, 2 << 30);
        b.piece_count = 100_000;
        let v = rank_candidates(vec![b.clone(), a.clone()], 1.0, 8);
        assert_eq!(v[0].piece_count, 100);
        // Negative bias is small-first by size; equal sizes keep both.
        let v = rank_candidates(vec![b, a], -1.0, 8);
        assert_eq!(v.len(), 2);
    }

    #[test]
    fn eviction_score_mirrors_rank() {
        // Ascending eviction order must be the exact reverse of descending
        // rank order for the same bias, so swaps reinforce ranking.
        let mut a = cand(1, 100 << 20);
        a.piece_count = 500;
        let mut b = cand(1, 900 << 20);
        b.piece_count = 500;
        for bias in [1.0, 0.5, -0.5, -1.0] {
            let ranked = rank_candidates(vec![a.clone(), b.clone()], bias, 8);
            let first = ranked[0].size_bytes;
            let sa = eviction_score(&a, bias, 8);
            let sb = eviction_score(&b, bias, 8);
            let evict_first = if sa < sb { a.size_bytes } else { b.size_bytes };
            assert_ne!(
                first, evict_first,
                "bias={bias}: rank-first ({first}) must be evicted last ({evict_first})"
            );
        }
    }

    #[test]
    fn bias_never_crosses_seeder_bands() {
        // A 10 TiB 1-seeder torrent outranks a 1 KiB 2-seeder one at any bias.
        let mut big = cand(1, 10 << 40);
        big.piece_count = 100;
        let mut small = cand(2, 1 << 10);
        small.piece_count = 100;
        for bias in [-1.0, 0.0, 1.0] {
            let v = rank_candidates(vec![small.clone(), big.clone()], bias, 8);
            assert_eq!(v[0].seeders, 1, "bias={bias}");
        }
    }

    #[test]
    fn chance_math() {
        // floor 0/1 -> aggressiveness^(seeders-1)
        assert!((selection_chance(0.6, 1, 0) - 1.0).abs() < 1e-12);
        assert!((selection_chance(0.6, 2, 0) - 0.6).abs() < 1e-12);
        // at/below floor -> 1.0
        assert!((selection_chance(0.6, 3, 5) - 1.0).abs() < 1e-12);
        assert!((selection_chance(0.6, 5, 5) - 1.0).abs() < 1e-12);
        assert!((selection_chance(0.6, 6, 5) - 0.6).abs() < 1e-12);
    }

    #[test]
    fn floor_p10() {
        assert_eq!(seeder_floor(&[]), 0);
        assert_eq!(seeder_floor(&[0, 0]), 0);
        // 10 values: rank ceil(10/10)=1 -> min
        assert_eq!(seeder_floor(&[1, 1, 2, 3, 4, 5, 6, 7, 8, 9]), 1);
        // 20 values 1..=20: rank 2 -> 2
        let v: Vec<u32> = (1..=20).collect();
        assert_eq!(seeder_floor(&v), 2);
    }

    #[test]
    fn swap_gates() {
        let c = cand(1, 10);
        let d = evaluate_swap(&c, &[], 2, 0.6, 0.5);
        assert!(d.should_swap);
        let d = evaluate_swap(&c, &[], 2, 0.6, 0.999);
        assert!(d.should_swap); // chance 1.0: any roll < 1 passes
        let c2 = cand(3, 10);
        let d = evaluate_swap(&c2, &[], 2, 0.6, 0.5);
        assert!(!d.should_swap);
        assert!(d.seed_scarcity_blocked());
        // margin
        let held = vec![Held {
            info_hash: [1u8; 20],
            title: String::new(),
            size_bytes: 5,
            piece_count: 0,
            seeders: 4,
        }];
        let d = evaluate_swap(&cand(3, 10), &held, 2, 0.6, 0.0);
        assert!(!d.should_swap);
        assert!(!d.seed_scarcity_blocked());
        let d = evaluate_swap(&cand(2, 10), &held, 2, 0.6, 0.0);
        assert!(d.should_swap);
    }

    #[test]
    fn location_choice_weighted() {
        // Only index 1 fits.
        assert_eq!(choose_location(&[10, 100, 5], &[50, 50, 50], 0.0), Some(1));
        assert_eq!(choose_location(&[10, 5], &[50, 50], 0.0), None);
        // Proportional: roll 0 picks first eligible after sort.
        assert_eq!(choose_location(&[100, 100], &[10, 10], 0.0), Some(0));
        assert_eq!(choose_location(&[100, 100], &[10, 10], 0.999), Some(1));
    }
}
